### PR TITLE
Fix Memory Issue GQA CPU Rotary

### PR DESCRIPTION
**Pull request by aciddelgado**
### Description
In GQA there was a memory issue which was best described by @edgchen1 [here](https://github.com/microsoft/onnxruntime/issues/22252#issuecomment-2384559255)

> here's the problematic code:
> 
> https://github.com/microsoft/onnxruntime/blob/d9de054eb53034e3dc18c298e47c6cc08d5aa884/onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc#L149-L157
> 
> annotated:
> 
> ```c++
>     if (packed_qkv) {
>       // Q is an OrtValue declared in the enclosing scope.
>       OrtValue RotaryQKV;
>       Tensor::InitOrtValue(element_type, TensorShape({batch_size, num_heads_ + 2 * kv_num_heads_, sequence_length, head_size}), allocator, RotaryQKV);
>       // Save pointer to Q's data in q_input.
>       q_input = Q.Get<Tensor>().Data<T>();
>       k_input = q_input + num_heads_ * sequence_length * head_size;
>       q_rotary = RotaryQKV.GetMutable<Tensor>()->MutableData<T>();
>       k_rotary = q_rotary + num_heads_ * sequence_length * head_size;
>       // Overwrite Q with RotaryQKV (OrtValues contain shared_ptr to contained value).
>       // Now, q_input is pointing to freed memory.
>       Q = RotaryQKV;
>     }
> ```
> 
> later on, when we use `q_input`, there is a read access violation.
> 
> https://github.com/microsoft/onnxruntime/blob/d9de054eb53034e3dc18c298e47c6cc08d5aa884/onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc#L170-L172
> 
> this problem showed up when CPU allocator sharing between sessions was enabled. in that case, the CPU allocator's arena was disabled. I suspect that the default usage of the arena hid this issue.
> 
> though I debugged into the first branch, this appears to be a problem in both branches:
> 
> https://github.com/microsoft/onnxruntime/blob/d9de054eb53034e3dc18c298e47c6cc08d5aa884/onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc#L149-L168

### Motivation and Context
Fixes a crucial bug. The issue was found here https://github.com/microsoft/onnxruntime/issues/22252

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc [hunk 1]
```diff
@@ -142,10 +142,14 @@ Status GroupQueryAttention<T>::Compute(OpKernelContext* context) const {
     const T* k_input;
     T* q_rotary;
     T* k_rotary;
+    // New OrtValues so data doesn't get freed
+    OrtValue Q_value_og;
+    OrtValue K_value_og;
     if (packed_qkv) {
       OrtValue RotaryQKV;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc:149]: The root cause is `OrtValue RotaryQKV` is declared within if scope. The memory will be released out of the if scope. The solution is to move `OrtValue RotaryQKV` to the outer scope and make sure it lives until kernel execution is finished.

I think the file has many place with similar issue. Please double check each OrtValue.

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc [hunk 2]
```diff
@@ -155,8 +159,10 @@ Status GroupQueryAttention<T>::Compute(OpKernelContext* context) const {
       Tensor::InitOrtValue(element_type, TensorShape({batch_size, num_heads_, sequence_length, head_size}), allocator, RotaryQ);
       OrtValue RotaryK;
       Tensor::InitOrtValue(element_type, TensorShape({batch_size, kv_num_heads_, sequence_length, head_size}), allocator, RotaryK);
-      q_input = Q.Get<Tensor>().Data<T>();
-      k_input = K.Get<Tensor>().Data<T>();
+      Q_value_og = Q;
+      K_value_og = K;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc:163]: I suggest to remove the following lines
```
Q = RotaryQKV;
...
Q = RotaryQ;
K = RotaryK;

Then change
```
  return ApplyAttention(Q.Get<Tensor>().Data<T>(), packed_qkv ? nullptr : K.Get<Tensor>().Data<T>(),
                        packed_qkv ? nullptr : V.Get<Tensor>().Data<T>(), past_key, past_value, output, present_k, present_v,
                        seqlens_k, parameters, allocator, context);
```
to
```
  return ApplyAttention(q_rotary, packed_qkv ? nullptr : k_rotary,
                        packed_qkv ? nullptr : V.Get<Tensor>().Data<T>(), past_key, past_value, output, present_k, present_v,
                        seqlens_k, parameters, allocator, context);
```

**Review by tianleiwu — APPROVED** [review]

**Review by edgchen1 — APPROVED** [review]

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc [hunk 3]
```diff
@@ -102,6 +102,11 @@ Status GroupQueryAttention<T>::Compute(OpKernelContext* context) const {
         allocator, batch_size, kv_num_heads_, sequence_length, head_size, value, V));
   }
 
+  OrtValue RotaryQKV;
+  OrtValue RotaryQ;
+  OrtValue RotaryK;
+  T* q_rotary = Q.GetMutable<Tensor>()->MutableData<T>();
+  T* k_rotary = K.GetMutable<Tensor>()->MutableData<T>();
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc:109]: T* k_rotary = packed_qkv ? nullptr : K.GetMutable<Tensor>()->MutableData<T>();

**Comment by skyline75489** [comment]
I can confirm this fixes the crash. Thanks!

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**